8189441: Define algorithm names for keys derived from KeyAgreement

wangweij opened the pull request:
Allow `Generic` as an algorithm in the `KeyAgreement::generateSecret(alg)` method.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8347059](https://bugs.openjdk.org/browse/JDK-8347059) to be approved

### Issues
 * [JDK-8189441](https://bugs.openjdk.org/browse/JDK-8189441): Define algorithm names for keys derived from KeyAgreement (**Bug** - P4)
 * [JDK-8347059](https://bugs.openjdk.org/browse/JDK-8347059): Define algorithm names for keys derived from KeyAgreement (**CSR**)


### Reviewers
 * [Sean Mullan](https://openjdk.org/census#mullan) (@seanjmullan - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22650/head:pull/22650` \
`$ git checkout pull/22650`

Update a local copy of the PR: \
`$ git checkout pull/22650` \
`$ git pull https://git.openjdk.org/jdk.git pull/22650/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22650`

View PR using the GUI difftool: \
`$ git pr show -t 22650`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22650.diff">https://git.openjdk.org/jdk/pull/22650.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22650#issuecomment-2528833736)
</details>

wangweij commented:
The comparison of algorithm names was inconsistent, sometimes `equals` and sometimes `equalsIgnoreCase`. This is even more suspicious in the `P11KeyAgreement` where both were used. My understanding is that `equalsIgnoreCase` should always be used because key algorithm has always been case-insensitive. Also, since the "TlsPremasterSecret" algorithm is only used internally, it is always provided with the correct case.

seanjmullan commented on src/java.base/share/classes/com/sun/crypto/provider/DHKeyAgreement.java:
```diff
@@ -419,10 +419,14 @@ protected SecretKey engineGenerateSecret(String algorithm)
             return skey;
-        } else if (algorithm.equals("TlsPremasterSecret")) {
+        } else if (KeyUtil.isSupportedKeyAgreementOutputAlgorithm(algorithm)) {
             // remove leading zero bytes per RFC 5246 Section 8.1.2
```
Maybe move this comment after line 424 inside the `if` block.

seanjmullan commented on test/jdk/java/security/KeyAgreement/Generic.java:
```diff
@@ -0,0 +1,68 @@
+ * @test
+ * @bug 8189441
+ * @library /test/lib /test/jdk/sun/security/pkcs11
+ * @summary make sure Generic is accepted by all KeyAgreement
```
Suggest: s/KeyAgreement/KeyAgreement implementations/

seanjmullan commented on src/java.base/share/classes/javax/crypto/KeyAgreement.java:
```diff
@@ -659,9 +659,16 @@ public final int generateSecret(byte[] sharedSecret, int offset)
-     * @param algorithm the requested secret-key algorithm
-     *
-     * @return the shared secret key
+     * @param algorithm the requested secret-key algorithm. This is
```
Did you consider instead referencing the new section you are adding for standard key algorithm names (https://bugs.openjdk.org/browse/JDK-8346736)? It could be reworded as:

"See the Key Algorithms section in the Java Security Standard Algorithm Names Specification for information about standard secret key algorithm names. Use "Generic" ..."

wangweij replied:
Good idea.

wangweij replied:
Do you suggest me adding the same to `encapsulate` and `decapsulate` of `KEM` in this PR?

seanjmullan replied:
Yes.

wangweij replied:
Finally, no. KEM changes will be in https://github.com/openjdk/jdk/pull/22929.

seanjmullan commented:
Can you also replace the link in `Key.getAlgorithm` with the new section you will be adding?

wangweij commented:
> Can you also replace the link in `Key.getAlgorithm` with the new section you will be adding?

OK, and maybe more. I just found `SecretKeySpec` and `EncodedKeySpec`.

wangweij commented:
I've modified too many files and I think they should belong to https://bugs.openjdk.org/browse/JDK-8346736. We should only touch KEM and KeyAgreement here.

seanjmullan commented:
> I've modified too many files and I think they should belong to https://bugs.openjdk.org/browse/JDK-8346736. We should only touch KEM and KeyAgreement here.

Makes sense.

wangweij commented:
I've reverted all changes not related to `KeyAgreement` and put them in https://github.com/openjdk/jdk/pull/22929, including the `KEM` one, since it involves no behavior change and is pure clarification. I've also removed the "KEM" word from the bug title.

seanjmullan commented on src/java.base/share/classes/javax/crypto/KeyAgreement.java:
```diff
@@ -659,9 +659,19 @@ public final int generateSecret(byte[] sharedSecret, int offset)
+     *      <a href="{@docRoot}/../specs/security/standard-names.html#secretkey-algorithms">
+     *      Java Security Standard Algorithm Names Specification</a>
+     *      for information about standard secret key algorithm names.
+     *      Use "Generic" if the output will be used as the input keying
```
s/Use/Specify/

seanjmullan commented on src/java.base/share/classes/javax/crypto/KeyAgreement.java:
```diff
@@ -659,9 +659,19 @@ public final int generateSecret(byte[] sharedSecret, int offset)
-     * @return the shared secret key
+     * @return the shared secret key. The length of the key material
+     *      may be adjusted to be compatible with the specified algorithm,
+     *      regardless of whether the key is extractable. If {@code algorithm}
+     *      is specified as "Generic", the full shared secret will be returned.
```
I think this last sentence is too strict, as some existing provider implementations may not support "Generic". I would suggest rewording as "If {@code algorithm} is specified as "Generic" and is supported by the implementation, the full shared secret is returned."

seanjmullan commented on src/java.base/share/classes/sun/security/ec/XDHKeyAgreement.java:
```diff
@@ -207,9 +209,9 @@ protected SecretKey engineGenerateSecret(String algorithm)
+        if (!KeyUtil.isSupportedKeyAgreementOutputAlgorithm(algorithm)) {
             throw new NoSuchAlgorithmException(
-                    "Only supported for algorithm TlsPremasterSecret");
+                    "Unsupported secret key algorithm " + algorithm);
```
Might be clearer to add a `:` after algorithm, also more consistent with DHKeyAgreement.

seanjmullan commented on src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11KeyAgreement.java:
```diff
@@ -268,20 +268,18 @@ protected SecretKey engineGenerateSecret(String algorithm)
-        if (algorithm.equals("TlsPremasterSecret")) {
+        if (KeyUtil.isSupportedKeyAgreementOutputAlgorithm(algorithm)) {
             // For now, only perform native derivation for TlsPremasterSecret
-            // as that is required for FIPS compliance.
+            // as that is required for FIPS compliance and Generic for KDF.
```
Suggest breaking into two sentences: "For now, only perform native derivation for TlsPremasterSecret and Generic algorithms. TlsPremasterSecret is required for FIPS compliance and Generic is required for input to KDF."

seanjmullan commented on src/java.base/share/classes/javax/crypto/KeyAgreement.java:
```diff
@@ -659,9 +659,20 @@ public final int generateSecret(byte[] sharedSecret, int offset)
+     *      may be adjusted to be compatible with the specified algorithm,
+     *      regardless of whether the key is extractable. If {@code algorithm}
+     *      is specified as "Generic" and it is supported by the implementation,
+     *      the full shared secret will be returned.
```
s/will be/is/

seanjmullan commented on src/java.base/share/classes/javax/crypto/KeyAgreementSpi.java:
```diff
@@ -205,18 +205,30 @@ protected abstract int engineGenerateSecret(byte[] sharedSecret,
+     *      may be adjusted to be compatible with the specified algorithm,
+     *      regardless of whether the key is extractable. If {@code algorithm}
+     *      is specified as "Generic" and it is supported by the implementation,
+     *      the full shared secret will be returned.
```
s/will be/is/

seanjmullan reviewed (commented):
A few more minor comments.

wangweij commented:
/integrate
